compiler/src/service/compile.rs

```diff
@@ -0,0 +1,564 @@
+
+            estimated_shadow_files: Default::default(),
+            latest_doc: None,
+
+            steal_send,
+            steal_recv,
+
+            memory_send,
+            memory_recv,
+        }
+    }
+
+    /// Run the compiler thread synchronously.
+    pub fn run(self) -> bool {
+        use tokio::runtime::Handle;
+
+        if Handle::try_current().is_err() && self.enable_watch {
+            log::error!("Typst compiler thread with watch enabled must be run in a tokio runtime");
+            return false;
+        }
+
+        tokio::task::block_in_place(move || Handle::current().block_on(self.block_run_inner()))
+    }
+
+    /// Inner function for `run`, it launches the compiler thread and blocks
+    /// until it exits.
+    async fn block_run_inner(mut self) -> bool {
+        if !self.enable_watch {
+            let compiled = self
+                .compiler
+                .with_stage_diag::<false, _>("compiling", |driver| driver.compile());
+            return compiled.is_some();
+        }
+
+        if let Some(h) = self.spawn().await {
+            // Note: this is blocking the current thread.
+            // Note: the block safety is ensured by `run` function.
+            h.join().unwrap();
+        }
+
+        true
+    }
+
+    /// Spawn the compiler thread.
+    pub async fn spawn(mut self) -> Option<JoinHandle<()>> {
+        if !self.enable_watch {
+            self.compiler
+                .with_stage_diag::<false, _>("compiling", |driver| driver.compile());
+            return None;
+        }
+
+        // Setup internal channels.
+        let (dep_tx, dep_rx) = tokio::sync::mpsc::unbounded_channel();
+        let (fs_tx, mut fs_rx) = tokio::sync::mpsc::unbounded_channel();
+
+        // Wrap sender to send compiler response.
+        let compiler_ack = move |res: CompilerResponse| match res {
+            CompilerResponse::Notify(msg) => dep_tx.send(msg).unwrap(),
+        };
+
+        // Spawn file system watcher.
+        tokio::spawn(super::watch_deps(dep_rx, move |event| {
+            fs_tx.send(event).unwrap();
+        }));
+
+        // Spawn compiler thread.
+        let compile_thread = ensure_single_thread("typst-compiler", async move {
+            log::debug!("CompileActor: initialized");
+
+            // Wait for first events.
+            while let Some(event) = tokio::select! {
+                Some(it) = fs_rx.recv() => Some(CompilerInterrupt::Fs(it)),
+                Some(it) = self.memory_recv.recv() => Some(CompilerInterrupt::Memory(it)),
+                Some(it) = self.steal_recv.recv() => Some(CompilerInterrupt::Task(it)),
+            } {
+                // Small step to warp the logical clock.
+                self.logical_tick += 1;
+
+                // Accumulate events.
+                let mut need_recompile = false;
+                need_recompile = self.process(event, &compiler_ack) || need_recompile;
+                while let Some(event) = fs_rx
```

Comment:
i'd prefer breaking nested code here into 3 whiles. But... it's not my code so do it as you wish